README.md

```diff
@@ -322,6 +322,10 @@ Perform initial distribution of buckets across replicasets.
 
```

Comment:
The main idea is that the function called by the router API with explicit buckets in the first argument can process only the data from these buckets. It is the user's responsibility (vshard acts only as a transport), otherwise we get data anomalies.

The main idea for the `router:multical(buckets, .., fn, args)` is that we don't want to call multiple times on the router the same function for each bucket to process the bucket's data. Many individual calls can be very expensive (the original case is a remote SQL call), so it is be better to unite these calls into a single per-storage call (if possible). If we fail to group the calls into a single per-storage call, it can be bad for performance reasons, but we still have to get the correct results. So, we assume that `fn(b1, b2)` returns the same result as independent chain of calling `fn(b1)` and `fn(b2)`.

With these assumptions I can finally answer the questions.

> Is 2.c done only if any buckets are missing in 2.b is false?

2.c is executed if at least one bucket was found on 2.b. For example, we have sent an SQL query 
```sql
select * from sharded_space where bucket_id in (1, 2)
```
to the buckets 1 and 2. At step 1 we decided that both buckets 1 and 2 live on the storage 1 (according to the router cache). On 2.b we have found out that bucket 1 is absent on the storage 1 (but the bucket 2 is still there). So, we remember bucket 1 and still execute 2.c as the function itself can process only the data corresponding to the buckets 1 and 2... and as bucket 1 is absent, it will process only the bucket 2 rows. When bucket 1 new storage would be found on the next iteration, we'll execute the same function again and finally get `fn(b2)` and `fn(b1)` results (that are equal to `fn(b1, b2)`).

> I don't understand the guarantees the function is given if it is executed. Are all buckets guaranteed to be there or not?
What happens if some storages had all needed buckets and some did not? Are the successful storages also retried? (If yes, then it shouldn't be so, because you would redo write-mode requests which are not guaranteed to be idempotent.)

Function responsibility:
* `fn(b1, b2) == fn(b1) || fn(b2)` 
* each bucket data should be processed by the function exactly once. **<-** I have understood right now that the function can't do it in the current implementation. Looks like the function must always consume the current bucket list from the `storage:multicall(b_ids, func, {b_ids, ...})` first argument to solve the problem.

Function guarantees:
* it would be executed on all the listed buckets

> Is the storage-side function executed as many times as there are buckets or just once per storage?

In ideal world the function should be executed once per storage, in the worst case - as much as there are buckets in the list. In reality the value would be somewhere above the amount of storages. And according to the function responsibilities, these numbers should not influence the final result.

> What if the timeout expired and I already called the function on some buckets, but not on others? What is the error object then? Just a timeout error?

Just a timeout error.

Reply:
> 2.c is executed if at least one bucket was found on 2.b

Doesn't seem to be true in the patch then:
```Lua
        -- Reference the bucket.
        ok_ref, err = bucket_ref(bucket_id, mode)
        if not ok_ref then
            err = lerror.make(err)
            goto fail
        end
```
this is done on the storage side now. If you fail to ref any of the buckets, then the whole request fails. Is that intentional?

> each bucket data should be processed by the function exactly once.
> `fn(b1, b2) == fn(b1) || fn(b2)`

This won't work in the current implementation. Assume you want to process buckets 100 and 200. They are on separate storages. You've sent the request to the two storages. But firstly it has reached the one having bucket 100. It was processed there. Now the bucket 100 migrates to the same storage where bucket 200 is located. Now your remaining call-request arrives to that storage too, and will process buckets {100, 200}, even though 100 was already seen before.

Reply:
> and will process buckets {100, 200}, even though 100 was already seen before

Looks like I need to keep references to the buckets until the whole query is done (something like vshard does for the map -reduce with storage references).

New algorithm:

1. Group buckets by replicasets based on the router's cache.
2. For each replicaset from 1:
a. Send function, arguments and replicaset buckets from 1 to the storage.
b. Try to refer the buckets. If any buckets are missing or unavailable to refer, remember them.
c. Execute the function (pass referred buckets from 2.b as an additional argument - add it as the last one to the argument list) and save the result.
d. In a case of error or a timeout expiration on 2.c unrefer the buckets from 2.b and return error.
e. In a case of success on 2.c **do not unrefer** the buckets and return results from 2.c with unavailable buckets from 2.b to the router.
3. Collect results from all replicate sets on the router. If there are no errors, but some unavailable buckets were returned, search for these buckets and update the router bucket cache through bucket_resolve. With updated routes of these buckets go to 1.
4. At this moment we should unrefer the buckets on all the storages that didn't return errors (storages with errors have unreferred the buckets themselves on 2.d). The returned unavailable buckets were not referred. For the referred buckets the routing cache contains correct replicasets. For each replicaset from 1:
a. Exclude unavailable buckets 2.b from the grouped buckets 1 (this is required only if any result from 3 contains an error).
b. Send a new unref_buckets function to the storage with buckets from 4.a. It should not fail (execute it).
5. Wait 4.b to finish and return result or error (if there was an error on step 3).

Reply:
This could work if not the fact that if the storage connection breaks, then you won't be able to unreference the buckets on step 4. They will stay referenced until restart or until manual intervention. Same will happen if the router itself simply restarts between steps 3 and 4. I am afraid, you can't avoid potential double visit of buckets unless you do a full featured map-reduce, which already exists anyway - it gives you all the guarantees that you want.

Speaking of 2.c and the bucket list as the last argument - this doesn't look good. Mandatory positional arguments should have a fixed position. I suggest to make it the first argument.

Then about the approach in general. If you really much want to avoid the double-bucket visit, then I can suggest to implement it as a partial map-reduce. So it works exactly like `map_callrw()` with storage-wide references with timeouts, but walks only the replicasets where the needed buckets are stored. Then you don't even need to pass the bucket list to the storages, because all the buckets on them will be valid for reading and writing anyway. To be frank, I think this should be good enough for you. If you do that, then firstly please propose the possible API how it would look.

A more complex solution would be to evolve map-reduce even further - introduce time-limited bucket refs or even bucket-pack refs (to handle many buckets at a time). Would work like storage-wide refs, but wouldn't affect not needed buckets. The problems here are that it requires careful design, preferably in text version firstly. And that you still need to make it work with the rebalancer so as those bucket-refs wouldn't block all the buckets on the storage and the rebalancer wouldn't be able to work.

Reply:
> I can suggest to implement it as a partial map-reduce

A partial map-reduce doesn't work for my case. At the starting position on the router I have an SQL query and can derive buckets from the filter conditions in the WHERE clause. But the translation of these buckets to the storages is based on the routing cache and can be outdated. So, I can miss some storages to visit to get correct results.

>  introduce time-limited bucket refs or even bucket-pack refs

Looks like time-limited bucket refs is my choice to fix the new algorithm (to refer the buckets in 2.b with a timeout and don't care about unref in a case of a router failure anymore). I'll think about the design a little.

Reply:
> But the translation of these buckets to the storages is based on the routing cache and can be outdated.

Well, just like with your current solution. And then you retry. The same could be done with a partial map-reduce. Retry it until all the buckets are covered or the timeout expires.

The time-limited bucket refs would be an ideal solution, but this is going to be really complicated.

Reply:
After some thinking I agree that on the first iteration it makes sense to use partial map-reduce (for complexity reasons). We can improve this API with bucket timeouts on the next step.

So, the new algorithm

(**Router** means that we switch to router and execute the steps on the router until we switch to the storage with **Storage**):

1. **Router:** Group buckets by replicasets based on the router's cache.
2. Ref stage. For each storage:
a. Async send ref id, timeout and a group of the corresponding buckets to the storage.
    The aim is to reference this storage and check what buckets are absent.
4. **Storage:** Refer session with the ref id and timeout, passed from the router.
5. Lookup for the passed buckets. If any of them were not found on the storage,
   return them back in response to the router.
6. **Router:** Await and collect returned responses. If timeout has expired, set the error for
   this response.
7. If any of responses contains error,send unref to the refed storages and return
   the error to the user.
8. If the collected results contain moved buckets, search for them and update
   the router's cache. Decrease the timeout and goto 1.
9. Map stage. For each storage:
   a. Execute a callback, that modifies original arguments for current storage based on 
      the grouped buckets from 1. Under normal conditions it simply replaces a bucket 
      list with a group of buckets refed on the target storage. But the real 
      implementations can be much more complicated.
   b. Async send a map function with modified arguments and a ref id to the storage.
10. **Storage:** Execute storage_map: if the ref id has expired, return error. Otherwise,
   ref.use -> execute -> ref.del (as usual). Return results.
11. **Router:** Await results. If timeout expired, return error to the user. Otherwise, return
    result.

Argument transformation callback: f(args, buckets_by_storage) -> new_args

Storage API: storage_ref_with_lookup(rid, timeout, buckets) -> moved_buckets

Router API: map_multicallrw(bucket_ids, func, args, callback, opts)

Reply:
This could work. I only don't get why do you need the "transformation callback". Simply give the actual bucket list to your function on the storage side as an argument.

Reply:
I agree, the callback is redundant. At first I planed to use it for internal optimixations (deserialize one of the arguments with logical SQL plan, remove redundant nodes and serialize back, but this idea has a horrible performance). So, let's remove it from the algorithm.

Storage API: storage_ref_with_lookup(rid, timeout, buckets) -> moved_buckets
Router API: map_multicallrw(bucket_ids, func, args, opts)

I'll push the new commits on the next week.